sstables/uncompress.cc

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (C) 2024-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#include <cstdint>
+#include <seastar/core/fstream.hh>
+
+#include "seastar/core/future.hh"
+#include "seastar/core/temporary_buffer.hh"
+
+#include "types.hh"
+#include "exceptions.hh"
+#include "checksum_utils.hh"
+
+namespace sstables {
+
+// File data source implementation for uncompressed SSTables
+template <ChecksumUtils ChecksumType>
+class uncompressed_file_data_source_impl : public data_source_impl {
+    std::optional<input_stream<char>> _input_stream;
+    const checksum& _checksum;
+    uint64_t _underlying_pos;
+    uint64_t _pos;
+    uint64_t _beg_pos;
+    uint64_t _end_pos;
+public:
+    uncompressed_file_data_source_impl(file f, const checksum& checksum,
+    uint64_t pos, size_t len, file_input_stream_options options)
+    : _checksum(checksum)
+    , _pos(pos)
+    , _beg_pos(pos)
+    , _end_pos(pos + len)
+    {
+        // _beg_pos and _end_pos specify positions in the stream.
+        // These are not necessarily aligned on chunk boundaries.
+        // To be able to verify the checksums, we need to translate
+        // them into a range of chunks that contain the given
+        // address range, and open a file input stream to read that
+        // range. The _underlying_pos always points to the current
+        // chunk-aligned position of the file input stream.
+        auto chunk_size = checksum.chunk_size;
+        SCYLLA_ASSERT(chunk_size);
```

Comment:
Better to use on_internal_error(), which doesn't take the server down but tries to make progress.